File: examples/ruby/tree_1.rb
class Node
  attr_accessor :val, :left, :right
  def initialize(val)
    @val = val
    @left = @right = nil
  end
end

def insert(root, val)
  return Node.new(val) unless root
  val < root.val ? root.left = insert(root.left, val) : root.right = insert(root.right, val)
  root
end

def inorder(root)
  return [] unless root
  inorder(root.left) + [root.val] + inorder(root.right)
end

root = nil
[4, 2, 6, 1, 3].each { |x| root = insert(root, x) }
puts inorder(root).inspect

File: examples/php/tree_1.php
<?php
class Node {
    public $val;
    public $left;
    public $right;
    function __construct($val) { $this->val = $val; $this->left = $this->right = null; }
}
function insert($root, $val) {
    if ($root === null) return new Node($val);
    if ($val < $root->val) $root->left = insert($root->left, $val);
    else $root->right = insert($root->right, $val);
    return $root;
}
function inorder($root) {
    if ($root === null) return [];
    return array_merge(inorder($root->left), [$root->val], inorder($root->right));
}
$root = null;
foreach ([4, 2, 6, 1, 3] as $v) $root = insert($root, $v);
echo implode(',', inorder($root)) . "\n";
